src/electron/office/document_client.cc

```diff
@@ -0,0 +1,393 @@
+  gfx::Rect dirty_rect_twips_ =
+      gfx::ScaleToEnclosedRect(rect, lok_callback::TWIP_PER_PX);
+  constexpr int scale_factor = 1;
+  constexpr int tile_size_px = 256;
+
+  int tile_size_px_scaled = tile_size_px * scale_factor;
+  int nTileSizePixelsScaledTwips =
+      tile_size_px_scaled * lok_callback::TWIP_PER_PX;
+  long doc_width_scaled = document_size_px_.width() * scale_factor;
+  long doc_height_scaled = document_size_px_.height() * scale_factor;
+
+  // Total number of rows / columns in this document.
+  uint rows =
+      ceil(static_cast<double>(doc_height_scaled) / tile_size_px_scaled);
+  uint columns =
+      ceil(static_cast<double>(doc_width_scaled) / tile_size_px_scaled);
+
+  SkCanvas canvas(image_data);
+
+  // Render the tiles.
+  for (uint row = 0; row < rows; ++row) {
+    for (uint column = 0; column < columns; ++column) {
+      gfx::Rect tile_rect_twips, tile_rect_px;
+
+      // Determine size of the tile: the rightmost/bottommost tiles may
+      // be smaller, and we need the size to decide if we need to repaint.
+      if (column == columns - 1)
+        tile_rect_px.set_width(doc_width_scaled - column * tile_size_px_scaled);
+      else
+        tile_rect_px.set_width(tile_size_px_scaled);
+      if (row == rows - 1)
+        tile_rect_px.set_height(doc_height_scaled - row * tile_size_px_scaled);
+      else
+        tile_rect_px.set_height(tile_size_px_scaled);
+
+      // Determine size and position of the tile in document coordinates,
+      // so we can decide if we can skip painting for partial rendering.
+      tile_rect_twips = gfx::Rect(
+          lok_callback::PixelToTwip(tile_size_px_scaled, zoom_) * column,
+          lok_callback::PixelToTwip(tile_size_px_scaled, zoom_) * row,
+          lok_callback::PixelToTwip(tile_rect_px.width(), zoom_),
+          lok_callback::PixelToTwip(tile_rect_px.height(), zoom_));
+
+      if (dirty_rect_twips_.Intersects(tile_rect_twips)) {
+        SkImageInfo image_info = SkImageInfo::Make(
+            tile_size_px_scaled, tile_size_px_scaled,
+            tile_mode_ == LOK_TILEMODE_BGRA ? kBGRA_8888_SkColorType
+                                            : kRGBA_8888_SkColorType,
+            kPremul_SkAlphaType);
+        SkBitmap tile;
+        uint8_t* pixels = static_cast<uint8_t*>(
+            calloc(tile_size_px_scaled * tile_size_px_scaled, 4));
+
+        document_->paintPartTile(
+            pixels, 0, tile_size_px_scaled, tile_size_px_scaled,
+            tile_rect_twips.x(), tile_rect_twips.y(),
+            nTileSizePixelsScaledTwips, nTileSizePixelsScaledTwips);
+
+        tile.installPixels(
+            image_info, pixels, image_info.minRowBytes(),
+            [](void* addr, void* context) {
+              if (addr)
+                free(addr);
+            },
+            nullptr);
+        canvas.drawImage(tile.asImage(), tile_size_px_scaled * column,
+                         tile_size_px_scaled * row);
+      }
+    }
+  }
```

Comment:
This can be optimized, but it's the mostly a port of the code from lokdocview.cxx

Basically we know the starting column: `floor(rect.left / tile_size_px)` and ending column `ceil(rect.right / tile_size_px)`.  
Similarly so for the row, but with `top` and `bottom` coordinates respectively. This could probably be calculated using `gfx::ScaleToEnclosingRect(rect, 1 / tilesize)`, then retrieved using `.x()`, `.right()`, `.y()`, `.bottom()`.